mlir/include/Quantum/IR/QuantumOps.td

```diff
@@ -224,25 +225,37 @@ def CustomOp : Gate_Op<"custom", [AttrSizedOperandSegments, DifferentiableGate]>
         Variadic<F64>:$params,
         Variadic<QubitType>:$in_qubits,
         StrAttr:$gate_name,
-        OptionalAttr<UnitAttr>:$adjoint
+        OptionalAttr<UnitAttr>:$adjoint,
+        Variadic<QubitType>:$in_ctrl_qubits,
+        Variadic<I1>:$in_ctrl_values
```

Comment:
This is just a point for discussion: 

I'm not a huge fan about how bloated this op is becoming, but unfortunately I don't necessarily know if there is a better and practical way :( 

A structured op like in your previous PR could be an alternative, but it does come with its own problems as well:
- potential overheads (like requiring a terminator and handling input/output values to the region)
- making the lowering of quantum operations depend on parent context (but maybe this isn't unusual?)
- are there others?

However, this form is also not ideal:
- these modifiers need to be added to all existing gates (we only have 3 but in principle we'd like to have more to better support quantum compilation passes)
- it's not really scalable to add more and more modifiers like this (you've talked about full on operator arithmetic before which does tie into this point)
- beefy operations with many different modes are harder to work with than simple operations that only do one thing (e.g. we can see changes in the passes and tests that require additional elements to be there)

What are your thoughts on this? (either way this PR should be kept simple, so if we use a different representation it should be limited like a structured operation with only one child op)

Reply:
> This is just a point for discussion:
> 
> I'm not a huge fan about how bloated this op is becoming, but unfortunately I don't necessarily know if there is a better and practical way :(
> 
> A structured op like in your previous PR could be an alternative, but it does come with its own problems as well:
>     * potential overheads (like requiring a terminator and handling input/output values to the region)
>     * making the lowering of quantum operations depend on parent context (but maybe this isn't unusual?)
>     * are there others?

The structured control operation from my previous PR would not save us from having this `ctrl_qubits` / `ctrl_values` fields in the individual operations. At that time we did not have nested quantum tapes enabling the frontend-side transformations, so I the idea was to use MLIR passes to transform controlled scopes into sequences of controlled operations. In order to encode the latter, I started adding individual ctrl fields in that PR as well (only did it for custom op).

Another option I can think of - is to keep ctrl regions, and then write a pass transforming regions with only one op to the LLVM API call directly, like `ctrl(q1,q2) { quantum.custom "PauilX"() q0; } --> __catalyst_qis_PauliX(q0, ctrl=[q1,q2])`. In this case no individual ctrl fields would be needed, but I think the ConversionPatterns sources would be unnecessary complicated.   

Anyways, now we can distribute regions in the frontend, so this PR includes only what we need to encode our device API calls.

> However, this form is also not ideal:
> 
>     * these modifiers need to be added to all existing gates (we only have 3 but in principle we'd like to have more to better support quantum compilation passes)

Hopefully we are not going to have more gates. Actually, as I already asked somewhere, I would suggest to merge MultiRZ gate into quantum.custom, so we have only two: quantum.custom (equivalent of a function) and quantum.unitary (equivalent of a numeric constant) in our quantum language. 

>     * it's not really scalable to add more and more modifiers like this (you've talked about full on operator arithmetic before which does tie into this point)

Indeed, setting `in_ctrl_qubits` field is a form of saying that we tensor-multiplying our quantum operator with another pre-defined operator of special form. I would love to have the full set of linear arithmetic (addition, multiplication by constant, tensor products) - all constexprs, known at the compile time - in our quantum dialect. If/when we decide to add these, I will be happy to deprecate the ctrl fields. 

>     * beefy operations with many different modes are harder to work with than simple operations that only do one thing (e.g. we can see changes in the passes and tests that require additional elements to be there)

Hmm, but compared with other approaches I though of this way might actually be a simple one. Consider the idea for which I posted a draft ADR sometimes ago - to encode the operator arithmetic using mlir operators. This way would require us to first parse mlir program to get the arithmetic structure, then do the transformations (and then print everything back as mlir, but it is simple) .

> 
> What are your thoughts on this? (either way this PR should be kept simple, so if we use a different representation it should be limited like a structured operation with only one child op)

Beyond the scope of this PR:
 
I was thinking on how to find a compromise between the locality (keeping the information about arithmetic in a single operation to simplify pass-based processing) with completeness (having multiplicaiton, addition, tensor products, not only control or adjoint). Just an idea: what about introducing a simple operator language with its own parser/printer as a string attribute of an operator? Might sound silly, but look

```  mlir
%b1, %b2, %b3 = quantum.custom "Kron(PauliX, Ctrl(1,PauliY))" %a1, %a2, %3   
```

In C++ we will need to have only a syntax checker for these strings (much much simpler than mlir passes), and only one final pass to translate this into the sequence of matrix calculations and device API calls. What do you think?

Reply:
I think "bloated" is subjective here. It reflects the PennyLane's lightning's API. (Whether this is a good thing or a bad thing, that's a different question). This reminds me of ISAs. This ISA would be something like: The first N bits encode the instruction, the next M bits are the target qubits, the next O bits are the control qubits, the next P bit defines whether the adjoint operation is taking place. Variable length ISAs.

Other ISAs, have something like: target register, operand registers, bits to encode instructions, and potentially some bit shifts depending on an immediate value.

To. me though, one thing that is closely related to decomposition is whether or not this control option or control qubits won't be decomposed by the frontend. In other words, would quantum ISAs support adding optional control qubits, or would it be expected for the compiler to decompose control(operation) into other equivalent operations? I still see control and adjoint as meta operations (operations that take operations and produce new operations). Whether that is true or not, I am not 100% sure.

Can we allocate some time to look at other quantum ISAs and see how difficult it would be to target them? The only description of an ISA for quantum computers I've seen is here: https://arxiv.org/pdf/1909.11719.pdf but I didn't look too much.

Reply:
@grwlf regarding the operator language: I'll think about it, thanks! Why is it simpler than extracting this information from MLIR? Why can't we explicitly define these operations in MLIR?

Reply:
> @grwlf regarding the operator language: I'll think about it, thanks! Why is it simpler than extracting this information from MLIR? Why can't we explicitly define these operations in MLIR?

For clarity, by MLIR encoding I mean something like this:
``` mlir
%op1 = quantum.constOp "PauliZ"
%op2 = quantum.constOp "PauliX"
%op2 = quantum.userOp {
   ^(q: !quantum.qubit):
     %q2 = quantum.call %op1 %q
     yield %q2
   }
%op3 = quantum.add %op1, %op2
%op4 = quantum.mul %op3, 3.14
```

In the above imaginary language a similar program might look like  `Mul(Add(PauliZ, PauliX), 3.14)`.

* I think that these quantum-expressions are simpler because I see them as something that we don't assign to MLIR variables. As a consequence, there is no need to define the behavior for these variables for other primitives, like loops or functions.
  * But this might become a downside - without variable there is no ways to link several string-expressions into one (here I assume that  `Mul(Add(PauliZ, %expr), 3.14)` is illegal).
* Of cause, we will need to write a string-parser, but I think that
  * It is still not hard, we can use common parser toolkits
  * As an alternative, maybe we can define a similar syntax using custom types of an MLIR dialect, like they do for affine maps, for example

Reply:
Thank you for these comments, we can take these into account when updating the IR to meet our optimization needs :)